### PR TITLE
Add RPC stream for tracking hosts in Sampi.

### PR DESCRIPTION
**Pull request by archseer**
Used for watching new hosts.

### CODE REVIEW

File: sampi/rpc.go [hunk 1]
```diff
@@ -93,3 +95,29 @@ func (s *Cluster) RemoveJobs(hostID *string, jobIDs []string, res *struct{}) err
...
+		s.state.RemoveListener(id, ch)
+		close(ch)
+	}()
+	for {
+		select {
+		case event := <-ch:
+			select {
+			case stream.Send <- event:
+			case <-stream.Error:
+				return nil
```
Comment (lmars) [inline sampi/rpc.go:117]: You should either return the error from `stream.Error`, or drop the return value of this method and just `break` here

File: sampi/state.go [hunk 2]
```diff
@@ -14,7 +14,9 @@ type State struct {
 	curr *map[string]host.Host
 	next *map[string]host.Host
 
-	streams map[string]chan<- *host.Job
+	listeners map[string]map[chan host.HostEvent]struct{} // host id -> listener list (ID "all" gets all events)
+	listenMtx sync.RWMutex
+	streams   map[string]chan<- *host.Job
```
Comment (lmars) [inline sampi/state.go:19]: What is this for?

File: sampi/rpc.go (hunk 1 above)
Comment (titanous) [inline sampi/rpc.go:117]: Actually, rpcplus returns it for us if we return nil.

Comment (lmars) [inline sampi/rpc.go:117]: Oh ok, so this method should have a return value of `error` even though it seemingly doesn't return one?

Comment (titanous) [inline sampi/rpc.go:117]: Yeah, RPC methods are special.

Comment (lmars) [inline sampi/rpc.go:117]: Ok I stand corrected :wink:

File: sampi/state.go [hunk 3]
```diff
@@ -131,3 +133,33 @@ func (s *State) RemoveHost(id string) {
 	delete(s.streams, id)
 	s.nextModified = true
 }
+
+func (s *State) AddListener(hostID string, ch chan host.HostEvent) {
```
Comment (titanous) [inline sampi/state.go]: I don't think we need to be able filter by host ID, it's relatively rare that we care about a specific host, and it adds additional complexity.

File: sampi/state.go (hunk 2 above)
Comment (archseer) [inline sampi/state.go:19]: I copied it over from the other client, it's how the listeners are stored for the event stream.

Comment (lmars) [inline sampi/state.go:19]: Sorry I missed the fact it was already there, I thought you had added it and not used it elsewhere, again ignore me :stuck_out_tongue:

File: sampi/state.go (hunk 3 above)
Comment (archseer) [inline sampi/state.go]: I was thinking we might add more events in the future, that would make listening for a single host more relevant, but you're probably right, I can't think of a use-case right now.

File: sampi/rpc.go [hunk 4]
```diff
@@ -81,6 +82,7 @@ outer:
 	s.state.Begin()
 	s.state.RemoveHost(h.ID)
 	s.state.Commit()
+	s.state.sendEvent(h, "remove")
```
Comment (titanous) [inline sampi/rpc.go]: This should happen in a goroutine (ditto for add), and pass just the ID:

``` go
go s.state.sendEvent(h.ID, "remove")
```

**Comment by lmars** [comment]
LGTM

File: sampi/rpc.go [hunk 5]
```diff
@@ -93,3 +95,29 @@ func (s *Cluster) RemoveJobs(hostID *string, jobIDs []string, res *struct{}) err
 	s.state.Commit()
 	return nil
 }
+
+func (s *Cluster) StreamHostEvents(stream rpcplus.Stream) error {
```
Comment (titanous) [inline sampi/rpc.go]: This needs a first argument (just make it a `struct{}`).

**Comment by titanous** [comment]
LGTM pending my comment. Squash please.

**Comment by archseer** [comment]
Squashed.

File: sampi/rpc.go [hunk 6]
```diff
@@ -93,3 +95,29 @@ func (s *Cluster) RemoveJobs(hostID *string, jobIDs []string, res *struct{}) err
 	s.state.Commit()
 	return nil
 }
+
+func (s *Cluster) StreamHostEvents(struct{}, stream rpcplus.Stream) error {
```
Comment (titanous) [inline sampi/rpc.go]: This is not valid Go, it needs a param name :)

**Comment by archseer** [comment]
_facedesks_

Really fixed this time :)

**Pull request merged**